sw/host/ot_certs/src/template/testgen.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides functionality to generate substitute data for template
+//! to test corner cases of the certificate generator.
+
+use anyhow::{ensure, Result};
+use rand::distributions::DistString;
+
+use openssl::bn::{BigNum, BigNumContext};
+use openssl::ec::{EcGroup, EcKey};
+use openssl::nid::Nid;
+use openssl::pkey::Private;
+
+use crate::template::subst::{SubstData, SubstValue};
+use crate::template::{EcCurve, EcPublicKeyInfo, SubjectPublicKeyInfo, Template, Value, Variable};
+
+// Convert a template curve name to an openssl one.
+fn ecgroup_from_curve(curve: &EcCurve) -> EcGroup {
+    match curve {
+        EcCurve::Prime256v1 => EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap(),
+    }
+}
+
+impl Template {
+    /// Generate a random set of data to substitute in the template to fill all the values.
+    /// This generator will make sure to generate valid values for the public key if possible.
+    pub fn random_test(&self) -> Result<SubstData> {
+        let mut data = SubstData::new();
+        for (var, var_type) in &self.variables {
+            let val = match *var_type {
+                super::VariableType::ByteArray { size } => {
+                    let bytes = (0..size).map(|_| rand::random::<u8>()).collect::<Vec<_>>();
+                    SubstValue::ByteArray(bytes)
+                }
+                super::VariableType::String { size } => {
+                    let s = rand::distributions::Alphanumeric
+                        .sample_string(&mut rand::thread_rng(), size);
+                    SubstValue::String(s)
+                }
+                super::VariableType::Integer { size } => {
+                    if size == 4 {
+                        // FIXME the code does not properly distinguish between signed and
```

Comment:
Is this going to be an issue, or fine to merge with this defect?

Reply:
It is fine to merge it now. I will fix it in a follow up since I want to rework the testgen to make it deterministic in the future.